src/main/java/com/github/manolo8/darkbot/core/objects/facades/AssemblyMediator.java

```diff
@@ -0,0 +1,276 @@
+
+        @Override
+        public void update() {
+            if (address == 0) return;
+            isCraftable = API.readBoolean(address + 0x20);
+
+            long itemVoAddress = Main.API.readMemoryPtr(address + 0x58);
+            recipeId = API.readMemoryString(itemVoAddress, 0x48);
+
+            long rewardsArrAddress = Main.API.readMemoryPtr(address + 0x60);
+            rewardsArr.update(rewardsArrAddress);
+            rewards.clear();
+            rewardsArr.forEach(ptr -> rewards.add(API.readMemoryString(ptr, 0x48)));
+
+            long resourcesRequiredArrAddress = Main.API.readMemoryPtr(address + 0x50);
+            resourcesRequiredArr.update(resourcesRequiredArrAddress);
+            resourcesRequiredArr.sync(resourcesRequired, ResourceRequired::new);
+        }
+
+        @Override
+        public String getRecipeId() {
+            return recipeId;
+        }
+
+        @Override
+        public List<? extends String> getRewards() {
+            return rewards;
+        }
+
+        @Override
+        public List<? extends AssemblyAPI.ResourceRequired> getResourcesRequired() {
+            return resourcesRequired;
+        }
+
+        @Override
+        public boolean isCraftable() {
+            return isCraftable;
+        }
+
+        @Override
+        public String toString() {
+            return Recipe.class.getSimpleName() + " - " + recipeId + " - " + isCraftable;
+        }
+    }
+
+
+    public static class ResourceRequired extends Auto implements AssemblyAPI.ResourceRequired {
+        public String resourceId = "";
+        public double amountRequired, amountRequiredBackup = -1.0;
+
+        public void update() {
+            if (address <= 0) return;
+            long itemVoAddress = Main.API.readMemoryPtr(address + 0x28);
+            resourceId = API.readMemoryString(itemVoAddress, 0x48);
+            amountRequired = API.readDouble(address + 0x30);
+            //this also gives back same value, not sure which is correct
+            amountRequiredBackup = API.readDouble(address + 0x38);
+        }
+
+        @Override
+        public String getResourceId() {
+            return resourceId;
+        }
+
+        @Override
+        public double getAmountRequired() {
+            return amountRequired;
+        }
+
+        @Override
+        public double getAmountRequiredBackup() {
+            return amountRequiredBackup;
+        }
+
+        @Override
+        public String toString() {
+            return ResourceRequired.class.getSimpleName() + " - " + resourceId + " - " + amountRequired;
+        }
+    }
+
+    public static class RowFilter extends Auto implements AssemblyAPI.RowFilter {
+        public ItemFilter first;
+        public ItemFilter second;
+
+        public void update() {
+            if (address <= 0) return;
+            long firstAddress = Main.API.readMemoryPtr(address + 0x20);
+            if (firstAddress > 0) {
+                first = new ItemFilter();
```

Comment:
You should be able to just declare these off the bat, and call update(addr) without needing to re-create them

Reply:
If the address doesnt exist, no need to create them. also Im using this for null check to add them in filters list

Reply:
issue here is you're going to create new item filters *every single tick*, and nothing is setting them back to null if they were to stop being there either (ie: if they address went to 0, you'd just keep the previous value). So the question is, what is it really you're getting here? i think not much.

The solution is simply to always initialize both, and add them to the bigger list of address is != 0. Also should implement syncAndReport instead of relying on size*2 for updating the real filters